### PR TITLE
Added a certain admin request

### PR DESCRIPTION
**Pull request by DeityLink**
image

### CODE REVIEW

**Comment by dunsel-hordon** [comment]
nice

**Comment by ShiftyRail** [comment]
Fuck no

File: code/datums/gamemode/gamemode.dm [hunk 1]
```diff
@@ -250,6 +251,10 @@
 		F.process()
 	for(var/datum/role/R in orphaned_roles)
 		R.process()
+	if (runescape_pvp)
+		for (var/entry in runescape_fighters)
+			var/datum/runescape_fighter_data/the_data = runescape_fighters[entry]
```
Comment (ShiftyRail) [inline code/datums/gamemode/gamemode.dm]: make this a proper role datum, there is absolutely no point or benefit in having two separate system, role datums are supposed to have a universal character

Comment (DeityLink) [reply code/datums/gamemode/gamemode.dm]: The point is to have it be something much cheaper and independent from role datums (don't want to clog the round end scoreboard either). It doesn't need any of the variables that role datums use.

Comment (DeityLink) [reply code/datums/gamemode/gamemode.dm]: At the very least I moved all the stuff in mob.dm to its own file under modules.

Comment (DeityLink) [reply code/datums/gamemode/gamemode.dm]: More importantly, we don't want it to interfere with other roles in any way. Some features interfere with players that already have a role and I don't want being flagged as a runescape fighter causing such interference.

Comment (ShiftyRail) [reply code/datums/gamemode/gamemode.dm]: Then those features are the problem. 
What are they even? I am sorry, but no, I am not letting this new type of role not use the role datums framework.

Comment (DeityLink) [reply code/datums/gamemode/gamemode.dm]: That's not a type of role, it's an bus-only toggle button that makes skulls appear on top of players who've attacked someone else (unless they're attacking someone who attacked them in the past 5 minutes). And also pacifies everyone standing outside of maint, this is all for the purpose of a bus that ancient has in mind. This won't become part of every rounds.

It's completely independant of roles. There are no objectives, no score tracking, etc. Heck, the datum even disappears along with the skull when the player dies or 20 MINUTES go by. There are more reasons to not have it be a role datum than to have it be one. And I say that as one of the guys who coded the bloody datums.

To compare, it's as if bombermen players in bomberman arenas needed to have their own role datum. it's unnecessary, and would clog the end round scoreboard.

File: code/modules/mob/living/living.dm [hunk 2]
```diff
@@ -1557,6 +1558,10 @@ Thanks.
 		to_chat(src, "<span class='warning'>You can't do that now!</span>")
 		return FAILED_THROW
 
+	if(runescape_pvp && is_pacified())
+		to_chat(src, "<span class='warning'>As such, throwing items is also forbidden outside of maint.</span>")
```
Comment (Kurfursten) [inline code/modules/mob/living/living.dm]: Informal "maint" shouldn't appear in system messages

also, "as such"?

**Comment by Kurfursten** [comment]
This doesn't actually need a CL as it is an admin secret, and the CL currently is so vague as to only cause confusion

File: code/modules/mob/living/living.dm (hunk 2 above)
Comment (DeityLink) [reply code/modules/mob/living/living.dm]: "As such" because is_pacified() already sends a message when trying to attack outside of maint when this mode is enabled.

Comment (DeityLink) [reply code/modules/mob/living/living.dm]:
```suggestion
		to_chat(src, "<span class='warning'>As such, throwing items is also forbidden outside of maintenance areas.</span>")
```

**Comment by ShiftyRail** [comment]
You say there's no score tracking but your runescape fighter datum has several vars. It processes things. It is clear that runescape fighters can only fight other fighters - a check made easier if they are in their own faction, rather than pacifying everyone and disabling this check outside of maint - pacify only membres of the runescape faction. This makes it easier to add something else to the round, notably a bomberman arena if some runescape fighters die.

That this roundtype is meant to be a very rare occurrence is even more damning. If it is rare, why does it need a separate system exclusive to it, littering the code and making it harder to understand how it works at a glance ?

Role, factions, and dynamic rulesets were made to be as generic as possible and I will not allow for another snowflake system. This is a huge step backward in term of code quality, maintainabilty, and setting a good example for future coders.

See my tag mode/changeling clown PR for an example of a "bus-only ruleset".

**Comment by ShiftyRail** [comment]
Sorry didn't meant to close it

**Comment by DeityLink** [comment]
> It is clear that runescape fighters can only fight other fighters

> pacify only membres of the runescape faction.

That is not the case at all. It's a server-wide toggle. Everyone would be part of the faction.

My point is that it's a completely different goal, it's not a role
it's just a tracker of "should that mob have a skull icon on top of their head"
heck it's not even tied to the mind but to the body
if the skull was only visible to mobs of a same faction I could understand
but here it's visible to all mobs
role datums exist to
* track antags, which doesn't apply here because the skull may appear whether players are antag or not
* allow admins to easily visualize and manage said antags, which once again isn't a concern here
* give and track objectives, of which none are generated here
* track all of that then display it on the round end scoreboard, which isn't the case here
* be easily managed by Dynamic Mode to balance threat in the round etc

Is it because I'm updating the skulls with the /gamemode/process that you think it should be a role?
having extra roles in the round would also interfere with midround/latejoin ruleset injection probabilities
(admins may or may not want to disable rulesets when doing this bus, but both should be supported)

And once again it's not a ruleset, it's something that admins can enable and disable on the fly. All data is discarded on purpose when turning it off to allow resetting the skulls.

And lastly, it is completely self-contained, it's not as if they were interacting with other systems like roles actually do.

File: code/datums/gamemode/gamemode.dm (hunk 1 above)
Comment (mrflarg) [reply code/datums/gamemode/gamemode.dm]: the french fighting each other is good

**Comment by PrimeDSS13** [comment]
having an icon over your head isn't a role silly

**Comment by PrimeDSS13** [comment]
this isn't a role or a mode shifty, it's just an icon over peoples heads that eventually vanishes

File: code/datums/gamemode/gamemode.dm (hunk 1 above)
Comment (Kammerjunk) [reply code/datums/gamemode/gamemode.dm]: >it's as if bombermen players in bomberman arenas needed to have their own role datum

damn, that would be pretty cool. a separate area for roles that aren't really relevant if you're just wondering "was it valid for that guy to kill me," but you could still see stats like which "mini-role" killed who, which bomberman won the bomberround, etc

**Comment by ShiftyRail** [comment]
No. 
Tracking kills, having a special effect on whether or not the player is in maint or not, automatically pacifying players, processing at regular intervals, all of this needs to be integrated in the role datums framework or else this framework might as well not exist. 

The global var which controls whether or not skulls apply on top of players' heads behaves very much like a gamemode.

You can't just add an ad-hoc check for your runescape PVP meme in the pacifying procs and have a system of `/datum/runescape_fighter`s independent of role datums when a `/datum/role/runescape_fighter` for players would do the same. NPCs can have an ad-hoc system but for players it is vital that we keep the same infrastructure that everything else in the game, from traitors to the ERT including Tag Mode (#28601).

This is a question of maintainability, setting up a good example, and making this actually work with the game. **I can't expect a new coder to design his pseudo-gamemode or antag or whatever following proper standards if this is merged.** Minor feature or not, if everything else uses role datums, **this too must use role datums**, or a faction to list the runescape fighters.

In other words, make runescape fighter status a role tied to the mind if the thing is a player, like pretty much everything else in the game does.

**Comment by ShiftyRail** [comment]
This may comes off as a bit more blunt than I expected. What I mean by this is that you lose absolutely nothing by making the scope of the PR a little greated and integrating this status effect as a proper gamemode/ruleset/faction/role. 

I am sorry to be this much of an ass about it but letting this be a snowflake status effect not integrated to the rest of the gamemode "flow" is terribly shortsighted and ensures nobody will ever want to do anything with this.

**Comment by DeityLink** [comment]
> /datum/runescape_fighter

No such thing, as I've said I've renamed them in the last commit to make it clear that it's about the skull icon, not the fighter. It's a status effect of status effect just like hallucinations are a status effect and we don't do custom roles for those.

If anything, religions should be a role datum.

**Comment by PrimeDSS13** [comment]
Could you add a secondary admin button to this that doesn't disable combat outside of maint?

**Comment by DeityLink** [comment]
> 
> 
> Could you add a secondary admin button to this that doesn't disable combat outside of maint?

It is done. Toggling skulls and maint pvp are now two separate buttons.
